doc: improve documentation for trivial text writing functions (#277534)
NixOS/nixpkgs: doc/build-helpers/trivial-build-helpers.chapter.md

This sets [`allowSubstitutes` to `false`](https://nixos.org/nix/manual/#adv-attr-allowSubstitutes), so only use `runCommandLocal` if you are certain the user will always have a builder for the `system` of the derivation. This should be true for most trivial use cases (e.g., just copying some files to a different location or adding symlinks) because there the `system` is usually the same as `builtins.currentSystem`.
:::

## `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, `writeShellScriptBin` {#trivial-builder-textwriting}

Nixpkgs provides the following functions for producing derivations which write text into the Nix store:  `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin`.

`writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin` are convenience functions over `writeTextFile`.

These are useful for creating files from Nix expressions, which may be scripts or non-executable text files.

> Each of these functions will cause a derivation to be realized.  When you coerce the result of each of these functions to a string, it will evaluate to the *store path* of this derivation.

Reviewer: These functions won't cause a derivation to be realized. They will produce derivations as a side effect :sweat_smile: 

For now, they apart from creating a derivation as a side-effect, they will only produce as "direct output" an attribute set. This attribute set will contain at least the attribute `outPath` which is a special attribute as far as Nix-language is concerned.

When an attribute set `a` is string-interpolated as in `${a}`, the machinery will check if `outPath` (which precisely that symbol name) exists as an attribute on `a`: if yes, the value of `outPath` is used. 

These derivations will then later be realized ("made real?", "fully instantiated?") by other build passes. However, at this stage, we know where the artifacts of the derivation will live once made real. This is the `outPath` attribute on a `mkDerivation` style function.

:::: {.warning}

> Some of these functions will put the resulting files within a directory inside the derivation output.  If you need to refer to the resulting files somewhere else in Nix code, remember to append the path to the file  For example:

Reviewer: I like this. I finally understand it. (Did not quite understand this before)

Author: Yeah.  I think this warning is appropriate to put here, because these functions are some of the first ones a new Nix user will start to use (to create shell scripts, etc), and although it's obvious to folks who've been using it a while, it's not obvious at all to someone using it for the first time what the string coercion of the results of these functions will be.

:::

## `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, `writeShellScriptBin` {#trivial-builder-text-writing}

Nixpkgs provides the following functions for producing derivations which write text into the Nix store:  `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin`.

`writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin` are convenience functions over `writeTextFile`.

These are useful for creating files from Nix expressions, which may be scripts or non-executable text files.

Each of these functions will cause a derivation to be produced.  When you coerce the result of each of these functions to a string, it will evaluate to the *store path* of this derivation.

:::: {.warning}

> Some of these functions will put the resulting files within a directory inside the derivation output.  If you need to refer to the resulting files somewhere else in Nix code, remember to append the path to the file  For example:

Reviewer 2: ```suggestion
Some of these functions will put the resulting files within a directory inside the derivation output.  If you need to refer to the resulting files somewhere else in Nix code, remember to append the path to the file. For example:
```

> ```nix
> 
> # if the derivation destination is a directory....
> 
> my-file = writeTextFile {
>   name = "my-file";
>   text = ''
>     Contents of File
>   '';
>   destination = "/share/my-file";
> }
> 
> # remember to tack on "/share/my-file" to the derivation path when
> # using it elsewhere.
> 
> writeShellScript "evaluate-my-file.sh" ''
>   cat ${my-file}/share/my-file
> '';
> ```

Reviewer 2: ````suggestion

If the derivation destination is a directory...
```nix

my-file = writeTextFile {
  name = "my-file";
  text = ''
    Contents of File
  '';
  destination = "/share/my-file";
}

writeShellScript "evaluate-my-file.sh" ''
  cat ${my-file}/share/my-file
'';
```
::: {.note}
Remember to tack on "/share/my-file" to the derivation path when using it elsewhere.
:::
````

We should remove prose from code-blocks.

This sets [`allowSubstitutes` to `false`](https://nixos.org/nix/manual/#adv-attr-allowSubstitutes), so only use `runCommandLocal` if you are certain the user will always have a builder for the `system` of the derivation. This should be true for most trivial use cases (e.g., just copying some files to a different location or adding symlinks) because there the `system` is usually the same as `builtins.currentSystem`.
:::

> ## `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, `writeShellScriptBin` {#trivial-builder-textwriting}

Author: That "#trivial-builder-writeText" section header does indeed still exist in the file at https://github.com/NixOS/nixpkgs/pull/277534/files#diff-8a106493610a86c515f3db384752ad3ae3ccfe7ec7eaa2b1bb50674e23c961e8R185 .  It doesn't point at *exactly* the same place, but it does get them to the trivial builders chapter and makes more sense in the big scheme of things to name the section headings after the functions I think.

> ## `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, `writeShellScriptBin` {#trivial-builder-textwriting}

Reviewer 3: Ah, my bad, I missed that!

`nixpkgs` provides a number of functions that produce derivations which write text into the Nix store.  These include `writeTextFile`, `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin`, each of which is documented below.

These are useful for creating files from Nix expressions, which may be scripts or non-executable text files, depending on which of the functions is used and the arguments it takes.

> The result of each of these functions will be a derivation.  When you coerce the resulting derivation to text, it will evaluate to the *store path*. Importantly, it will not include the destination subpath produced by the particular function.  So, for example, given the following expression:

Reviewer: ```suggestion
Evaluation of a `trivialBuilder` function will return return an attribute set that can be [string interpolated](https://nixos.org/manual/nix/stable/language/string-interpolation.html?highlight=string%20inter#string-interpolation) as a directory path where the the derivation (produced as a side-effect of evaluation) will be realized to. 

Importantly, it this interpolation will not include the destination subpath produced by the particular function.  So, for example, given the following expression:
```

I am not sure to what extent we ought to be describing how derivations work here. Perhaps we can provide a link to the nix reference manual: https://nixos.org/manual/nix/stable/language/derivations.html

But in any case, In Nix lingo:

* functions (more generally, expressions) are "evaluated"
* derivations are "realized" 
* derivations produce attribute sets, which contain an `outPath` value used when interpolating them into strings

I don't know how to explain this well though. So perhaps you can massage my suggested sentence into being clearer, if you find it useful?

```nix

my-file = writeTextFile {
  name = "my-file";
  text = ''
    Contents of File
  '';
  destination = "/share/my-file";
}
```

If `my-file` is coerced to text, it will resolve to `/nix/store/<store path>`, like any other derivation.  It will *not* evaluate to `/nix/store/<store path>/share/my-file`.  So to use it elsewhere, as an example (in this case, within a shell script you're writing in a Nix expression via `writeShellScript`), you might do:

```nix
writeShellScript "evaluate-my-file.sh" ''
  cat ${my-file}/share/my-file
'';
```

This will produce the desired result.  However, the following will not, it will fail because the store path is a directory, and is not the `my-file` file.

```nix
writeShellScript "evaluate-my-file.sh" ''
  cat ${my-file}
'';

> ```

Reviewer 3: I think this whole explanation can be shortened:
- There's no need to show examples that will fail. The general expectation when people see a code example in the manual is that it describes something that works. If you want to display an example of something that doesn't work, put it inside a warning admonition, but I think it's not needed here.
- There's no need to educate the reader that derivations can be coerced to store paths and the details on that. This is outside the scope of this section of the manual.

My suggestion is to reword this in a way that reads more like:
"Some of these functions will put the resulting files in a directory inside the derivation output. If you need to refer to the resulting files somewhere else in Nix code, remember to append the path to the file, like in the following example:
```
```nix
my-file = writeTextFile {
  name = "my-file";
  text = ''Contents of File';
  destination = "/share/my-file";
}

# Somewhere else in the code. Note how we had to append /share/my-file to the path!
writeShellScript "evaluate-my-file.sh" ''
  cat ${my-file}/share/my-file
'';
```

See [this tutorial](https://nix.dev/tutorials/nix-language#derivations) for more details on how a derivation is coerced into a store path."

  executable = true;
  destination = "/bin/my-file";
}
```
:::

> 

Reviewer 2: Please make these into separate examples. Also, pull the prose out into the surrounding text.

Author: This has been done.

  executable = true;
  destination = "/bin/my-file";
}
```
:::

::: {.note}
The commands `writeText`, `writeTextDir`, `writeScript`, `writeScriptBin`, `writeShellScript`, and `writeShellScriptBin` documented below are convenience functions that wrap `writeTextFile`.
:::

### `writeText` {#trivial-builder-writeText}

Writes a text file to the store

`writeText` takes two arguments: `name` and `text`, each of which should be
a string.

`name`

: the name used in the Nix store path.

`text`

: will be the contents of the file.

> The store path will include the the name, and it will be a file. Any path separators and shell-reserved elements in the name are escaped to create the store path identifier.

Reviewer: What is meant by "it will be a file" in "The store path will include the name, and it will be a file."?

Edit, reading further examples like this down below, I guess I understand that this is trying to communicate: the `outPath` of the derivation is a file. (vs. the `outPath` is a directory).

Not sure if there's a clearer way to say this, because for some reason, it was confusing to me initially. Just an FYI; not important.

Here is an example.

::: {.example #ex-writeText}
# Usage of `writeText`
```nix
# Writes my-file to /nix/store/<store path>
writeText "my-file"
  ''
  Contents of File
  '';
```
:::

> This example is a simpler way to spell:

Reviewer: ```suggestion
This example is equivalent to:
```

(Idea: avoiding colloquialisms, for those who might not have English as a culturally-familiar language.)

(Provided as a suggestion only in case you find it okay to apply as is.)

```nix
writeTextFile {
  name = "my-file";
  text = ''
    Contents of File
  '';
}
```

### `writeTextDir` {#trivial-builder-writeTextDir}

Writes a text file within a subdirectory of the store.

`writeTextDir` takes two arguments: `path` and `text`, each of which should be a string.

`path`

: the destination within the Nix store path under which to create the file.

`text`

: the contents of the file.

> The store path will be a directory. The Nix store identifier will be generated based on various elements of the path.

Reviewer: Is the second sentence here necessary? "The Nix store identifier will be generated based on various elements of the path."

::: {.example #ex-writeTextDir}
# Usage of `writeTextDir`
```nix
# Writes contents of file to /nix/store/<store path>/share/my-file
writeTextDir "share/my-file"
  ''
  Contents of File
  '';
```
:::

> The example is a simpler way to spell:

Reviewer: ```suggestion
This example is equivalent to:
```

(Idea: avoiding colloquialisms, for those who might not have English as a culturally-familiar language.)

(Provided as a suggestion only in case you find it okay to apply as is.)

```nix
writeTextFile {
  name = "my-file";
  text = ''
    Contents of File
  '';
  destination = "share/my-file";
}
```

### `writeScript` {#trivial-builder-writeScript}

Writes a script within the store.

`writeScript` takes two arguments: `name` and `text`, each of which should be a string.

`name`

: the name used in the Nix store path.

`text`

: the contents of the file.

The created file is marked as executable.

The store path will include the the name, and it will be a file. Any path separators and shell-reserved elements in the name are escaped to create the store path identifier.

Here is an example.

::: {.example #ex-writeScript}
# Usage of `writeScript`
```nix
# Writes my-file to /nix/store/<store path> and makes executable
writeScript "my-file"
  ''
  Contents of File
  '';
```
:::

> The example is a simpler way to spell:

Reviewer: ```suggestion
This example is equivalent to:
```

(Idea: avoiding colloquialisms, for those who might not have English as a culturally-familiar language.)

(Provided as a suggestion only in case you find it okay to apply as is.)

```nix
writeTextFile {
  name = "my-file";
  text = ''
    Contents of File
  '';
  executable = true;
}
```

### `writeScriptBin` {#trivial-builder-writeScriptBin}

Writes a script within a "bin" subirectory of a subdirectory of the store.

`writeScriptBin` takes two arguments: `name` and `text`, each of which should be a string.

`name`

: the name used in the Nix store path and within the file generated under the store path.

`text`

: the contents of the file.

The created file is marked as executable.

The file's contents will be put into `/nix/store/<store path>/bin/<name>`.

The store path will include the the name, and it will be a directory. Any path separators and shell-reserved elements in the name are escaped to create the store path identifier.

::: {.example #ex-writeScriptBin}
# Usage of `writeScriptBin`
```nix
writeScriptBin "my-script"
  ''
  echo "hi"
  '';
```
:::

> The example is a simpler way to spell:

Reviewer: ```suggestion
This example is equivalent to:
```

(Idea: avoiding colloquialisms, for those who might not have English as a culturally-familiar language.)

(Provided as a suggestion only in case you find it okay to apply as is.)

```nix
writeTextFile {
  name = "my-script";
  text = ''
    echo "hi"
  '';
  executable = true;
  destination = "bin/my-script"
}
```

### `writeShellScript` {#trivial-builder-writeShellScript}

Writes a shell script to the store.

`writeShellScript` takes two arguments: `name` and `text`, each of which should be a string.

`name`

: the name used in the Nix store path.

`text`

: the contents of the file.

The created file is marked as executable.

This function is almost exactly like `writeScript`, but it prepends a shebang line that points to the runtime shell (usually bash) at the top of the file contents.

The store path will include the the name, and it will be a file. Any path separators and shell-reserved elements in the name are escaped to create the store path identifier.

Here is an example.

::: {.example #ex-writeShellScript}
# Usage of `writeShellScript`
```nix
writeShellScript "my-script"
  ''
  echo "hi"
  '';
```
:::

> The example is a simpler way to spell:

Reviewer: ```suggestion
This example is equivalent to:
```

(Idea: avoiding colloquialisms, for those who might not have English as a culturally-familiar language.)

(Provided as a suggestion only in case you find it okay to apply as is.)

```nix
writeTextFile {
  name = "my-script";
  text = ''
    #! ${pkgs.runtimeShell}
    echo "hi"
  '';
  executable = true;
}
```
### `writeShellScriptBin` {#trivial-builder-writeShellScriptBin}

Writes a shell script to a "bin" subdirectory of subdirectory of the store.

`writeShellScriptBin` takes two arguments: `name` and `text`, each of which should be a string.

`name`

: the name used in the Nix store path and within the file generated under the store path.

`text`

: the contents of the file.

This function is almost exactly like `writeScriptBin`, but it prepends a shebang line that points to the runtime shell (usually bash) at the top of the file contents.

The file's contents will be put into `/nix/store/<store path>/bin/<name>`.

The store path will include the the name, and it will be a directory. Any path separators and shell-reserved elements in the name are escaped to create the store path identifier.

::: {.example #ex-writeShellScriptBin}
# Usage of `writeShellScriptBin`
```nix
writeShellScriptBin "my-script"
  ''
  echo "hi"
  '';
```
:::

> The example is a simpler way to spell:

Reviewer: ```suggestion
This example is equivalent to:
```

(Idea: avoiding colloquialisms, for those who might not have English as a culturally-familiar language.)

(Provided as a suggestion only in case you find it okay to apply as is.)
